File: c-files/problem_40/problem40.c
/*
 *
 * Champernowne's constant
 *
 * Problem 40
 *
 * An irrational decimal fraction is created by concatenating the positive integers:
 *
 * 0.123456789101112131415161718192021...
 *
 * It can be seen that the 12th digit of the fractional part is 1.
 *
 * If dn represents the nth digit of the fractional part, find the value of the following expression.
 *
 * d1 x d10 x d100 x d1000 x d10000 x d100000 x d1000000
 *
 * problem40.c
 *
 *  Created on: Jan 23, 2014
 *      Author: jcyescas
 */

#include <stdio.h>
#include <string.h>

int main() {
  printf("Champernowne's constant");

  int size = 1000000;
  char fraction[size + 1]; //
  fraction[0] = '\0';

  int i;
  for (i = 1; i <= size; i++) {
    char number[8];
    sprintf(number, "%d", i);
    if ((strlen(number) + strlen(fraction)) > size) {
      break;
    }

    strcat(fraction, number);
  }

  //printf("\n%s", fraction);
  int total = 1;
  for (i = 1; i < size; i*= 10) {
    printf("\n%c", fraction[i - 1]);
    total *= fraction[i - 1] - '0';
  }

  printf("\nThe value of the expression is %d", total);

  return 0;
}

File: java-files/Problem40.java
package org.algorithms.euler;

/**
 * Champernowne's constant
 *
 * Problem 40
 * An irrational decimal fraction is created by concatenating the positive integers:
 *
 * 0.123456789101112131415161718192021...
 *
 * It can be seen that the 12th digit of the fractional part is 1.
 *
 * If dn represents the nth digit of the fractional part, find the value of the following expression.
 *
 * d1 x d10 x d100 x d1000 x d10000 x d100000 x d1000000
 *
 */
public class Problem40 {

  public static void main(String[] args) {
    StringBuffer sb = new StringBuffer();
    int limit = 1000000;
    for (int i = 1; i <= limit; i++) {
      sb.append(i);
    }

    int total = 1;
    for (int i = 1; i <= limit; i*= 10) {
      System.out.println(sb.charAt(i - 1));
      total *= Integer.parseInt(sb.charAt(i - 1) + "");
    }

    System.out.println("Total: " + total);
  }

  public static void main1(String[] args) {
    String sb = new String();
    int limit = 1000000;
    for (int i = 1; i <= limit; i++) {
      sb += "" + i;
    }

    int total = 1;
    for (int i = 1; i <= limit; i*= 10) {
      System.out.println(sb.charAt(i - 1));
      total *= Integer.parseInt(sb.charAt(i - 1) + "");
    }

    System.out.println("Total: " + total);
  }
}
